library/nist_kw.c

```diff
@@ -0,0 +1,650 @@
+}
+
+int mbedtls_nist_kw_setkey( mbedtls_nist_kw_context *ctx,
+                       mbedtls_cipher_id_t cipher,
+                       const unsigned char *key,
+                       unsigned int keybits,
+                       const int isWrap )
+{
+    int ret;
+    const mbedtls_cipher_info_t *cipher_info;
+
+    /*
+     * SP800-38F currently defines AES cipher as the only block cipher allowed, but:
+     * "For KW and KWP, the underlying block cipher shall be approved, and the block size shall be
+     *  128 bits. Currently, the AES block cipher, with key lengths of 128, 192, or 256 bits, is the only
+     *  block cipher that fits this profile."
+     */
+    if( cipher != MBEDTLS_CIPHER_ID_AES )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
+    cipher_info = mbedtls_cipher_info_from_values( cipher, keybits, MBEDTLS_MODE_ECB );
+    if( cipher_info == NULL )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
+    if( cipher_info->block_size != 16 )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
+    mbedtls_cipher_free( &ctx->cipher_ctx );
+
+    if( ( ret = mbedtls_cipher_setup( &ctx->cipher_ctx, cipher_info ) ) != 0 )
+        return( ret );
+
+    if( ( ret = mbedtls_cipher_setkey( &ctx->cipher_ctx, key, keybits,
+                               isWrap ? MBEDTLS_ENCRYPT : MBEDTLS_DECRYPT ) ) != 0 )
+    {
+        return( ret );
+    }
+
+    return( 0 );
+}
+
+/*
+ * Free context
+ */
+void mbedtls_nist_kw_free( mbedtls_nist_kw_context *ctx )
+{
+    mbedtls_cipher_free( &ctx->cipher_ctx );
+    mbedtls_platform_zeroize( ctx, sizeof( mbedtls_nist_kw_context ) );
+}
+
+/*
+ * KW / KWP - AE
+ */
+int mbedtls_nist_kw_wrap( mbedtls_nist_kw_context *ctx, int mode,
+                     const unsigned char *input, size_t in_len,
+                     unsigned char *output, size_t* out_len )
+{
+    int ret = 0;
+    size_t semiBlocks = 0;
+    size_t s ;
+    size_t olen, i, padlen = 0;
+    uint64_t t = 0;
+    unsigned char outBuff[ KW_SEMIBLOCK_LENGTH * 2 ];
+    unsigned char inBuff[ KW_SEMIBLOCK_LENGTH * 2 ];
+    unsigned char* R2 = output + KW_SEMIBLOCK_LENGTH;
+    unsigned char* A = output ;
+    const uint32_t  ICV1[] = NIST_KW_ICV1;
+    const uint32_t  ICV2[] = NIST_KW_ICV2;
+
+/*
+ * generate the String to work on
+ */
+    if( mode == MBEDTLS_KW_MODE_KW )
+    {
+/*
+ * According to SP800-38F Table 1, plaintext length size limits
+ * for KW is 2 to 2^54-1 semiblocks
+ */
+        if( ( in_len < 8 )
+#if SIZE_MAX > UINT_MAX
+                ||( in_len > 0x1FFFFFFFFFFFFF8 )
+#endif
+           )
+        {
+            ret = MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+            goto cleanup;
+        }
+
+
+        memcpy( output, ICV1, KW_SEMIBLOCK_LENGTH );
+        memcpy( output + KW_SEMIBLOCK_LENGTH, input, in_len );
+    }
+    else
+    {
+        /*
+         * According to SP800-38F Table 1, plaintext length size limits
+         * for KWP is 1 to 2^32-1 octets
```

Comment:
```diff
- plaintext length size limits for KWP is 1 to 2^32-1 octets
+ the plaintext length for KWP is must be between 1 and 2^32-1 octets inclusive.
```

Please double check that my suggested revision is correct.

Reply:
The numbers are correct. “is must be” → “must be” for grammar.